packages/backend/src/open_payments/payment/outgoing/service.ts

```diff
@@ -159,6 +185,326 @@ async function createOutgoingPayment(
+  limits: AccessLimits
+): boolean {
+  if (
+    limits.sendAmount &&
+    (limits.sendAmount.assetCode !== payment.asset.code ||
+      limits.sendAmount.assetScale !== payment.asset.scale)
+  ) {
+    return false
+  }
+  return (
+    !limits.receiveAmount ||
+    (limits.receiveAmount.assetCode === payment.receiveAmount.assetCode &&
+      limits.receiveAmount.assetScale === payment.receiveAmount.assetScale)
+  )
+}
+
+function estimateTotalAvailableAmounts<T extends GrantAccess>(
+  accessArray: T[],
+  payment: OutgoingPayment,
+  slippage: number
+) {
+  const availableSendAmount = accessArray.reduce(
+    (prev: bigint, access: T) =>
+      prev + (access.limits?.sendAmount?.value ?? BigInt(0)),
+    BigInt(0)
+  )
+
+  const availableReceiveAmount = accessArray.reduce(
+    (prev: bigint, access: T) =>
+      prev + (access.limits?.receiveAmount?.value ?? BigInt(0)),
+    BigInt(0)
+  )
+
+  const estTotalAvailableSendAmount =
+    availableSendAmount +
+    BigInt(
+      Math.floor(
+        (Number(payment.quote.lowEstimatedExchangeRate.b) /
+          Number(payment.quote.lowEstimatedExchangeRate.a)) *
+          (1 + slippage) *
+          Number(availableReceiveAmount)
+      )
+    )
+  const estTotalAvailableReceiveAmount =
+    availableReceiveAmount +
+    BigInt(
+      Math.floor(
+        (Number(payment.quote.lowEstimatedExchangeRate.a) /
+          Number(payment.quote.lowEstimatedExchangeRate.b)) *
+          (1 + slippage) *
+          Number(availableSendAmount)
+      )
+    )
+  return { estTotalAvailableSendAmount, estTotalAvailableReceiveAmount }
+}
+
+interface PaymentAccess extends GrantAccess {
+  paymentInterval?: Interval
+}
+
+// "payment" is locked by the "deps.knex" transaction.
+async function validateGrant(
+  deps: ServiceDependencies,
+  payment: OutgoingPayment,
+  grant: Grant
+): Promise<boolean> {
+  const grantAccess = grant.getAccess({
+    type: AccessType.OutgoingPayment,
+    action: AccessAction.Create
+  })
+
+  if (!grantAccess) {
+    // log?
+    return false
+  }
+
+  // Find grant access(es) that authorize this payment (pending send/receive limits).
+  const paymentAccess: PaymentAccess[] = []
+  const unlimitedAccess: GrantAccess[] = []
+  const unrelatedAccess: GrantAccess[] = []
+
+  for (const access of grantAccess) {
+    let noSendAmount = false
+    let noReceiveAmount = false
+    if (
+      validateAccess({
+        access,
+        payment,
+        identifier: `${deps.publicHost}/${payment.accountId}`
+      })
+    ) {
+      if (!access.limits) {
+        return true
+      }
+      if (!access.limits.sendAmount) {
+        noSendAmount = true
+      }
+      if (!access.limits.receiveAmount) {
+        noReceiveAmount = true
+      }
+      if (noSendAmount && noReceiveAmount) {
+        return true
+      }
+      let paymentInterval: Interval | undefined
+      if (access.interval) {
+        paymentInterval = getInterval(access.interval, payment.createdAt)
```

Comment:
Sounds like a good idea especially when we have very long repetitions.

Reply:
I decided to pass the interval to `validateAccess` rather than returning it. I had a hard time renaming `validateAccess` to make it clear that it validates access and returns an interval.

Reply:
I'm thinking `getInterval` should be called as little as possible since it has a linear search for a particular interval.
We can change `unlimitedAccess` and `unrelatedAccess` to both also be `PaymentAccess[]` to avoid re-calling `getInterval` in the loops below.
We _could_ also check access `limit`s before ever calling `getInterval`, so that `getInterval` potentially never called (as opposed to currently being called for every grant `access`). Maybe we do something like replace the first call to `validatePayment` with `validatePaymentLimits` and change the subsequent calls to `validatePaymentAccess`?

Reply:
You are right. I'll change it to use `getInterval` as little as possible.